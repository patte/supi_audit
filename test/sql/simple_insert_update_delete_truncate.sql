begin;

    create table public.members(
        id int primary key,
        name text not null
    );

    select audit.enable_tracking('public.members');

    insert into public.members(id, name)
    values (1, 'foo');

    update public.members
        set name = 'bar'
        where id = 1;

    delete from public.members;

    truncate table public.members;

    with remap as (
        select distinct on (u.id)
            u.id,
            row_number() over () stable_id
        from
            audit.record_version arv,
            unnest(array[arv.record_id, arv.old_record_id]) u(id)
        order by
            u.id asc
    )
    select
        row_number() over (order by arv.id) as row_id,
        r.stable_id as remapped_record_id,
        ro.stable_id as remapped_old_record_id,
        op,
        table_schema,
        table_name,
        record,
        old_record
    from
        audit.record_version arv
        left join remap r
            on arv.record_id = r.id
        left join remap ro
            on arv.old_record_id = ro.id
    order by
        row_id;
rollback;
