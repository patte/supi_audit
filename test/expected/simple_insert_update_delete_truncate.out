begin;
    create table public.members(
        id int primary key,
        name text not null
    );
    select audit.enable_tracking('public.members');
 enable_tracking 
-----------------
 
(1 row)

    insert into public.members(id, name)
    values (1, 'foo');
    update public.members
        set name = 'bar'
        where id = 1;
    delete from public.members;
    truncate table public.members;
    with remap as (
        select distinct on (u.id)
            u.id,
            row_number() over () stable_id
        from
            audit.record_version arv,
            unnest(array[arv.record_id, arv.old_record_id]) u(id)
        order by
            u.id asc
    )
    select
        row_number() over (order by arv.id) as row_id,
        r.stable_id as remapped_record_id,
        ro.stable_id as remapped_old_record_id,
        op,
        table_schema,
        table_name,
        record,
        old_record
    from
        audit.record_version arv
        left join remap r
            on arv.record_id = r.id
        left join remap ro
            on arv.old_record_id = ro.id
    order by
        row_id;
 row_id | remapped_record_id | remapped_old_record_id |    op    | table_schema | table_name |          record          |        old_record        
--------+--------------------+------------------------+----------+--------------+------------+--------------------------+--------------------------
      1 |                  3 |                        | INSERT   | public       | members    | {"id": 1, "name": "foo"} | 
      2 |                  3 |                      3 | UPDATE   | public       | members    | {"id": 1, "name": "bar"} | {"id": 1, "name": "foo"}
      3 |                    |                      3 | DELETE   | public       | members    |                          | {"id": 1, "name": "bar"}
      4 |                    |                        | TRUNCATE | public       | members    |                          | 
(4 rows)

rollback;
